File: 2779-maximum-beauty-of-an-array-after-applying-operation/2779-maximum-beauty-of-an-array-after-applying-operation.py
class Solution:
    def maximumBeauty(self, nums: List[int], k: int) -> int:
        nums.sort()
        n = len(nums)
        if n <= 1:
            return n
        def canit(target):
            stack = deque([[nums[0], 1]])
            curr = 1
            for num in nums[1:]:
                if num - stack[0][0] <= 2 * k:
                    if num > stack[-1][0]:
                        stack.append([num, 1])
                    elif num == stack[-1][0]:
                        stack[-1][1] += 1
                else:
                    while stack and num - 2 * k > stack[0][0]:
                        value, count = stack.popleft()
                        curr -= count
                curr += 1
                if curr >= target:
                    return True
            return False
        l = 0
        r = n
        while l <= r:
            mid = (l + r) // 2
            val = canit(mid)
            if val: 
                l = mid + 1
            else:
                r = mid - 1
        return r

File: 2779-maximum-beauty-of-an-array-after-applying-operation/2779-maximum-beauty-of-an-array-after-applying-operation.go
func maximumBeauty(nums []int, k int) int {
    sort.Ints(nums)
    var left int
    var result int = 1
    var n int = len(nums)
    for i := 1; i < n; i ++ {
        num := nums[i]
        for (left < n) && (nums[left] + 2* k < num) {
            left ++
        }
        result = max(result, i - left + 1)
    }
    return result
}
